gateways/braintree/request_builders.go

```diff
@@ -3,27 +3,47 @@ package braintree
 import (
 	"fmt"
 	"github.com/BoltApp/sleet"
+	"github.com/BoltApp/sleet/common"
+	braintree_go "github.com/braintree-go/braintree-go"
 )
 
-func buildAuthRequest(authRequest *sleet.AuthorizationRequest) *TransactionRequest {
+func buildAuthRequest(authRequest *sleet.AuthorizationRequest) (*braintree_go.TransactionRequest, error) {
 	billingAddress := authRequest.BillingAddress
 	card := authRequest.CreditCard
-	return &TransactionRequest{
-		Type:   TransactionTypeSale,
-		Amount: sleet.AmountToString(&authRequest.Amount),
-		CreditCard: &CreditCard{
+	amount, err := convertToBraintreeDecimal(authRequest.Amount.Amount, authRequest.Amount.Currency)
+	if err != nil {
+		return nil, err
+	}
+
+	request := &braintree_go.TransactionRequest{
+		Type:   "sale",
+		Amount: amount,
+		CreditCard: &braintree_go.CreditCard{
 			Number:         card.Number,
 			ExpirationDate: fmt.Sprintf("%02d/%02d", card.ExpirationMonth, card.ExpirationMonth%100),
 			CVV:            card.CVV,
 		},
-		BillingAddress: &Address{
+	}
+
+	if billingAddress != nil {
+		request.BillingAddress = &braintree_go.Address{
 			FirstName:         authRequest.CreditCard.FirstName,
 			LastName:          authRequest.CreditCard.LastName,
-			StreetAddress:     billingAddress.StreetAddress1,
-			Locality:          billingAddress.Locality,
-			Region:            billingAddress.RegionCode,
-			PostalCode:        billingAddress.PostalCode,
-			CountryCodeAlpha2: billingAddress.CountryCode,
-		},
+			StreetAddress:     common.SafeStr(billingAddress.StreetAddress1),
+			Locality:          common.SafeStr(billingAddress.Locality),
+			Region:            common.SafeStr(billingAddress.RegionCode),
+			PostalCode:        common.SafeStr(billingAddress.PostalCode),
+			CountryCodeAlpha2: common.SafeStr(billingAddress.CountryCode),
+		}
+	}
+	return request, nil
+}
+
+func convertToBraintreeDecimal(amount int64, currencyCode string) (*braintree_go.Decimal, error) {
```

Comment:
This seems gateway agnostic, so probably move this to a common helper function?

Reply:
No like this is specifically to the braintree.Decimal class - not general decimals?

Reply:
Aah, I see, my bad